File: shuting_ye_hw2/shuting_ye_task2.py
#! /usr/bin/env python3
# -*- coding:utf-8 -*-

import sys
from pyspark import SparkContext
from datetime import datetime

def myHash(i,j,B):
    """i, j are translated integer for each item"""
    return (-i+j) % B

def comb2(n):
    return (n-1)*n//2

def genTup(x):
    """x[0] and x[1] are already ordered"""
    item = tuple(set(x[1])-set(x[0]))
    idx = list(x[1]).index(item[0])+1
    return x[0][:idx]+item+x[0][idx:]


def in_freqbuck(x, bucksize, bitmap):
    hv = myHash(x[0],x[1],bucksize)
    return bitmap[hv//31] & (1<<(hv%31))

def genLi(ci, sup, part):
    """find truely frequent iemset"""
    return part.flatMap(lambda x:[(i,None) for i in ci if set(i)<=x]).groupByKey().filter(lambda x: len(x[1])>=sup).keys()

def genCi(prevl, k, bitmap=None, bucksize=0):
    """
    prevl: rdd of truely freq. itemset of size k-1
    bitmap: useful to filter the c2 
    return : itemset is a rdd of list of tuples
    a candidate must satify that all its immediate subset are freq. itemset
        which means its combination must have size of k(k-1)/2
    """
    prevl4j = prevl.map(lambda x:(None,x))
    itemset = prevl4j.join(prevl4j).map(lambda x:x[1]).filter(lambda x: x[0]<x[1])
    if k>2:
        return itemset.filter(lambda x: k-2==len(set(x[0])&set(x[1]))).coalesce(1).map(genTup).map(lambda x:(x,None)).groupByKey().filter(lambda x: comb2(k)==len(x[1])).keys()
    else:
        # pair in c2 must also hash to a freq. bucket
        return itemset.filter(lambda x: in_freqbuck(x,bucksize,bitmap)).coalesce(1)

   
def PCY_Alg(part, sup):
    """
    input: 
        part: rdd of one sample of the entire rdd
    return:
        list of all sizes L (rdd)
    """

    # pass one: output frequent singletons and frequent buckets
    allL=[]
    bucksize = 26131
    bitmap = [0]*((bucksize+30)//31)

    c1 = part.flatMap(lambda x:x[1])
    l1 = c1.map(lambda x:(x,None)).groupByKey().filter(lambda x: len(x[1])>=sup).keys()
    c1_li = sorted(set(c1.collect()))
    prevl = l1.map(lambda x: c1_li.index(x))
    allL.append(l1.map(lambda x:(x,)))

    # freq. buckets
    users = part.keys().collect()
    # renumber the items, str to integer
    translated_part = part.map(lambda x:(users.index(x[0]),set([c1_li.index(i) for i in x[1]])))
    flattp = translated_part.flatMapValues(lambda x: x)
    freqbuck = flattp.join(flattp).values().filter(lambda x:x[0]<x[1])\
        .map(lambda x: myHash(x[0],x[1],bucksize)).map(lambda x:(x,None)).groupByKey()\
        .filter(lambda x: len(x[1])>=sup).keys().collect()
    for i in freqbuck:
        bitmap[i//31] |= 1<<(i%31) # set the bit to 1


    k = 1
    while True:
        k+=1
        ci=[]
        if k==2:
            ci = genCi(prevl, k, bitmap=bitmap, bucksize=bucksize)
        else:
            ci = genCi(prevl, k)

        ci_li=ci.collect()
        li = genLi(ci_li, sup, translated_part.values())
        len_li = len(li.collect())
        if len_li==0:
            break

        allL.append(li.map(lambda x: tuple([c1_li[i] for i in x])))
        prevl = li


    return allL


def SON_Alg(baskets, sup, eptRDD):
    """
    First Map: 
        input: one partition of baskets, sup = sup/numPartition
        output: (F,None) where F can be one of all sizes of freq. itemset
    Fist Reduce:
        eliminate the duplicates
    Second Map:
        local count for each freq. itemsets
    Second Reduce:
        only output itemsets whose count >= sup 
    """
    numPartition = 2
    # numPartition = baskets.getNumPartitions()
    fraction_sup = sup//numPartition

    # pass1 
    sample = baskets.randomSplit([1]*numPartition)
    maxl,allL=0,[]
    for part in sample:
        partL = PCY_Alg(part, fraction_sup)
        if len(partL)>maxl:
            maxl=len(partL)
        allL.append(partL)
    
    candidates=[]
    for i in range(maxl):
        rddLi = eptRDD
        for np in range(numPartition):
            try:
                rddLi = rddLi.union(allL[np][i])
            except IndexError:
                continue
        candidates.append(sorted(rddLi.map(lambda x: (x,None)).groupByKey().keys().collect()))


    # pass2
    freqItemsets = []
    for can in candidates:
        fitemi = baskets.flatMap(lambda x:[(i,None) for i in can if set(i)<=x[1]]).groupByKey().filter(lambda x: len(x[1])>=sup).keys().collect()
        if len(fitemi)>0:
            freqItemsets.append(sorted(fitemi))

    return (candidates, freqItemsets)


def case1(rdd, thre, sup, eptRDD):

    # create baskets for each user, where business ids are unique
    user_basks = rdd.filter(lambda x: 'user_id' not in x).map(lambda x:x.split(','))\
        .groupByKey().map(lambda x: (x[0],set(x[1]))).filter(lambda x: len(x[1])>thre).persist()

    return SON_Alg(user_basks, sup, eptRDD)    


def task2(argv):
    thre, sup, infile, outfile = int(argv[1]), int(argv[2]), argv[3], argv[4]

    sc = SparkContext('local[*]','task2')

    start = datetime.now()
    ubRDD = sc.textFile(infile)
    eptRDD = sc.parallelize([])

    candidates, freqItemsets=case1(ubRDD, thre, sup, eptRDD)

    with open(outfile,'w') as outf:
        outf.write('Candidates:')
        for can in candidates:
            if can==candidates[0]:
                outf.write('\n'+','.join(str(c)[:-2]+str(c)[-1] for c in can))
            else:
                outf.write('\n\n'+','.join(str(c) for c in can))

        outf.write('\n\nFrequent Itemsets:')
        for fi in freqItemsets:
            if fi==freqItemsets[0]:
                outf.write('\n'+','.join(str(f)[:-2]+str(f)[-1] for f in fi))
            else:
                outf.write('\n\n'+','.join(str(f) for f in fi))


    end = datetime.now()
    print('Duration:', (end-start).seconds)

if __name__ == '__main__':
    task2(sys.argv)

File: shuting_ye_hw2/shuting_ye_task2.scala
import java.io._
import java.util.Date

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession


object shuting_ye_task2 {

  def myHash(i: Int, j: Int, B: Int): Int = (-i+j) % B

  def comb2(k: Int):Int = (k-1)*k/2

  /* Compare two ordered list of same size, act just like tuple comparision in python*/
  def mylt(a:List[Int], b:List[Int]):Boolean = {
    for (i<-0 until a.length){
      if (a(i) < b(i)) return true
      else if (a(i) > b(i)) return false
    }
    false
  }

  implicit val myordering = new Ordering[List[String]] {
    override def compare(x: List[String], y: List[String]): Int = {
      var ret = 0
      for (i <- 0 until x.length){
        ret = x(i).compareTo(y(i))
        if (ret!=0) return ret
      }
      ret
    }
  }

  def in_freqbuck(x: (Set[Int], Set[Int]), bucksize: Int, bitmap: Array[Int]): Boolean = {
    val hv = myHash(x._1.head,x._2.head,bucksize)
    (bitmap(hv/31) & (1<<(hv%31)))>0
  }

  def genLi(ci: Array[Set[Int]], sup: Int, part: RDD[Set[Int]]) =  part.flatMap(x=> for(i <- ci if i.subsetOf(x)) yield (i,1)).groupByKey().filter(x=> x._2.size>=sup).keys

  /* Especially to generate C2*/
  def genCi(prevl: RDD[Set[Int]], k: Int, bitmap: Array[Int], bucksize: Int):RDD[Set[Int]] = {
    val prevl4j = prevl.map(x=>(1,x))
    val itemset = prevl4j.join(prevl4j).values.filter(x=> x._1.head<x._2.head)
    itemset.filter(x=> in_freqbuck(x,bucksize,bitmap))
      .map(x=>x._1.++(x._2)).coalesce(1)
  }

  def genCi(prevl: RDD[Set[Int]], k: Int):RDD[Set[Int]] = {
    val prevl4j = prevl.map(x=>(1,x))
    val itemset = prevl4j.join(prevl4j).values.filter(x=> mylt(x._1.toList.sorted,x._2.toList.sorted))
    itemset.filter(x=> k-2==(x._1.&(x._2)).size).coalesce(1)
      .map(x=>(x._1.++(x._2),null)).groupByKey().filter(x=> comb2(k)==x._2.size).keys
  }


  def PCY_Alg(part: RDD[(String, Set[String])], sup: Int):List[RDD[Set[String]]] = {

    var allL = List[RDD[Set[String]]]()
    val bucksize = 26131
    var bitmap = Array.fill((bucksize+30)/31)(0)

    val c1 = part.flatMap(x=>x._2)
    val l1 = c1.map(x=>(x,null)).groupByKey().filter(x=> x._2.size>=sup).keys
    val c1_li = c1.collect().distinct.sorted
    var prevl = l1.map(x=> Set(c1_li.indexOf(x)))
    allL = allL:+(l1.map(x=>Set(x)))

    // freq. buckets
    val users = part.keys.collect()
    // renumber the items, str to integer
    val translated_part = part.map(x=>(users.indexOf(x._1), x._2.map(i=>c1_li.indexOf(i)) ))
    val flattp = translated_part.flatMapValues(x=>x)
    val freqbuck = flattp.join(flattp).values.filter(x=>x._1<x._2)
      .map(x=> myHash(x._1,x._2,bucksize)).map(x=>(x,null)).groupByKey()
      .filter(x=> x._2.size>=sup).keys.collect()

    for (i <- freqbuck) {
      bitmap(i/31) |= 1 << (i%31)
    }

    var k = 1
    var flag = true
    while(flag){
      k+=1
      val ci = if(k==2) genCi(prevl, k, bitmap, bucksize) else genCi(prevl, k)
      val ci_li=ci.collect()

      val li = genLi(ci_li, sup, translated_part.values)
      val len_li = li.collect().length
      allL = allL:+(li.map(x=> x.map(i=>c1_li(i)) ))
      prevl = li
      if (len_li==0) flag=false

    }

    allL
  }

  def SON_Alg(rdd: RDD[String], thre: Int, sup: Int): (Array[List[List[String]]], List[List[List[String]]]) = {

    val baskets = rdd.filter(x => !x.contains("user_id")).map(x => x.split(",")).map(x => (x(0),x(1)))
      .groupByKey().map(x => (x._1, x._2.toSet)).filter(x => x._2.size >thre).persist()

    val numPartition = 2
    val fraction_sup = sup / numPartition

    // pass1
    val sample = baskets.randomSplit(Array.fill(numPartition)(1.0))
    var maxli = 0
    val allL = new Array[List[RDD[Set[String]]]](numPartition)
    for(i <- 0 until sample.length){
      val partL = PCY_Alg(sample(i), fraction_sup)
      allL(i) = partL.dropRight(1)
      if(partL.length-1>allL(maxli).length) maxli = i
    }

    val maxl = allL(maxli).length
    val candidates = new Array[List[List[String]]](maxl)
    val zipres = allL(0).zip(allL(1))
    for (i<-0 until zipres.length){
      candidates(i) = zipres(i)._1.union(zipres(i)._2).map(x=>(x,null)).groupByKey().keys
        .map(x=> x.toList.sorted).collect().toList.sorted(myordering)
    }
    for (i<-zipres.length until maxl if maxl>zipres.length){
      candidates(i) = allL(maxli)(i).map(x=> x.toList.sorted).collect().toList.sorted(myordering)
    }

    // pass2
    var freqItemsets = List[List[List[String]]]()
    for(i<-0 until candidates.length){
      val fitemi = baskets.flatMap(x => for (c <- candidates(i) if c.toSet.subsetOf(x._2)) yield (c, null))
        .groupByKey().filter(x=> x._2.size>=sup).keys.collect()

      if(fitemi.length>0){
        freqItemsets = freqItemsets:+(fitemi.toList.sorted(myordering))
      }
    }

    (candidates, freqItemsets)
  }


  def main(args: Array[String]): Unit = {

    val thre = args(0).toInt
    val sup = args(1).toInt
    val infile = args(2)
    val outfile = args(3)

    val ss = SparkSession.builder().appName("task2").config("spark.master","local[*]")
      .getOrCreate()
    val sc = ss.sparkContext
    sc.setLogLevel("WARN")

    val start = new Date().getTime
    val ubRDD = sc.textFile(infile)

    val (candidates, freqItemsets) = SON_Alg(ubRDD, thre, sup)

    val outf = new FileWriter(new File(outfile))
    outf.write("Candidates:")
    for (can <- candidates){
      outf.write("\n" + can.map(x=> x.mkString("('", "', '", "')")).mkString(",") + "\n")
    }
    outf.write("\nFrequent Itemsets:")
    for (i <- 0 until freqItemsets.length){
      if (i<freqItemsets.length-1)
        outf.write("\n" + freqItemsets(i).map(x=> x.mkString("('", "', '", "')")).mkString(",")+"\n")
      else
        outf.write("\n" + freqItemsets(i).map(x=> x.mkString("('", "', '", "')")).mkString(","))
    }
    outf.close()

    val end = new Date().getTime
    println("Duration:"+(end-start)/1000)
  }

}
